server/src/main/java/org/opensearch/indices/replication/SegmentReplicator.java

```diff
@@ -102,6 +110,92 @@ SegmentReplicationTarget startReplication(
         return target;
     }
 
+    public ReplicationStats getSegmentReplicationStats(final ShardId shardId) {
+        final ConcurrentNavigableMap<Long, ReplicationCheckpointStats> existingCheckpointStats = replicationCheckpointStats.get(shardId);
+        if (existingCheckpointStats == null || existingCheckpointStats.isEmpty()) {
+            return ReplicationStats.empty();
+        }
+
+        Map.Entry<Long, ReplicationCheckpointStats> lowestEntry = existingCheckpointStats.firstEntry();
+        Map.Entry<Long, ReplicationCheckpointStats> highestEntry = existingCheckpointStats.lastEntry();
+
+        long bytesBehind = highestEntry.getValue().getBytesBehind();
+        long replicationLag = bytesBehind > 0L
+            ? TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - lowestEntry.getValue().getTimestamp())
+            : 0;
+
+        return new ReplicationStats(bytesBehind, bytesBehind, replicationLag);
+    }
+
+    public void updateReplicationCheckpointStats(final ReplicationCheckpoint latestCheckPoint, final IndexShard indexShard) {
+        latestReceivedCheckpoint.put(indexShard.shardId(), latestCheckPoint);
+        ConcurrentNavigableMap<Long, ReplicationCheckpointStats> existingCheckpointStats = replicationCheckpointStats.computeIfAbsent(
+            indexShard.shardId(),
+            k -> new ConcurrentSkipListMap<>()
+        );
+
+        long segmentInfosVersion = latestCheckPoint.getSegmentInfosVersion();
+        if (!existingCheckpointStats.containsKey(segmentInfosVersion)) {
```

Comment:
lets use atomic `.computeIfAbsent` here to prevent race conditions on ordering with `getSegmentReplicationStats` and `pruneCheckpointsUpToLastSync`.  We should also synchronize around these shard level maps in those methods to preserve high/low view.